### PR TITLE
Update manifest merger

### PR DESCRIPTION
**Pull request by marcinkwiatkowski**
This switches to `ManifestMerger2` implementation and fixes https://github.com/facebook/buck/issues/299

Majority of this PR is updating aosp to `25.2.0` version of the tools.

The only interesting change is in `GenerateManifestStep.java`.

For the sake of not maintaining two versions of aosp, this PR replaces old `ManifestMerger` with `ManifestMerger2`. Having both is possible (we could take `ManifestMerger` from tools `24.x.x`, the last version that shipped with it) but if the new merger works for everyone, probably best to just go with that.

### CODE REVIEW

**Comment by kageiit** [comment]
ManifestMerger2 also has a lot of improvements we contributed to aosp, most notably the fix to remove duplication of library manifest loading https://code.google.com/p/android/issues/detail?id=222034#c7

Excited to see this change in buck finally

**Comment by sdwilsh** [comment]
@dreiss or @marcinkosiba should probably be the ones to look at this.

**Comment by dreiss** [comment]
Can you rebase this?  I fixed some of the annotations that should not have been referencing the Android versions.

Can you also update the readme with enough detail for me to ensure that the steps can be reproduced?

I'll also need to check that this doesn't break any of our existing manifest merging at FB.

**Comment by marcinkwiatkowski** [comment]
@dreiss updated.

**Comment by Coneko** [comment]
`GenerateManifestStepTest` is failing both on Travis and on Appveyor, seems related to the change.

File: build.xml [hunk 1]
```diff
@@ -317,7 +317,7 @@
     <ecj srcdir="${aosp.src.dir}" destdir="${aosp.classes.dir}" gensrcdir="${aosp.gensrc.dir}">
       <ecj-params>
         <!-- ManifestMerger.java has redundant null checks -->
-        <compilerarg value="-err:-null" />
```
Comment (dreiss) [inline build.xml]: Why did this change?

File: third-party/java/aosp/README.txt [hunk 2]
```diff
@@ -1,13 +1,16 @@
...
+com.android.tools:sdklib:25.2.0
+com.android.tools.layoutlib:layoutlib-api:25.2.0
+2) Copied com.android.manifmerger.ManifestMerger2 and all required dependencies to the aosp/ module.
 3) In the Android SDK's source code manifmerger, sdklib and utils are separate modules. These are
 copied into a single module which alters certain package paths. For example,
-com.android.SDKConstants becomes com.android.common.SDKConstants.
-4) Added support for re-ordering <activity-alias> so they always go after <activity> elements.
-5) Added support for merging <meta-data> elements of <application>.
\ No newline at end of file
+com.android.SDKConstants becomes com.android.common.SDKConstants. This is done to avoid conflicts
```
Comment (dreiss) [inline third-party/java/aosp/README.txt]: Renaming SDKConstants is gross.  What libraries does this conflict with?  Should we update them?  This hack is fine for now, but you should file a task to clean it up.

File: third-party/java/aosp/aosp.iml [hunk 3]
```diff
@@ -17,6 +17,6 @@
         <SOURCES />
       </library>
     </orderEntry>
+    <orderEntry type="library" name="buck-lib" level="project" />
```
Comment (dreiss) [inline third-party/java/aosp/aosp.iml]: Why is buck-lib in here?

File: src/com/facebook/buck/android/GenerateManifestStep.java [hunk 4]
```diff
@@ -86,35 +84,48 @@ public StepExecutionResult execute(ExecutionContext context) {
...
-        String xmlText = Files.toString(outManifestFile, Charsets.UTF_8);
-        xmlText = xmlText.replace("\r\n", "\n");
-        Files.write(xmlText.getBytes(Charsets.UTF_8), outManifestFile);
-      } catch (IOException e) {
-        throw new HumanReadableException("Error converting line endings of manifest file");
-      }
+      xmlText = xmlText.replace("\r\n", "\n");
+    }
+    try {
+      File outManifestFile = outManifestPath.toFile();
```
Comment (dreiss) [inline src/com/facebook/buck/android/GenerateManifestStep.java]: Can you use ProjectFileSystem.writeContentsToPath here?

File: build.xml (hunk 1 above)
Comment (marcinkwiatkowski) [inline build.xml]: There is a ton of new warnings in `ManifestMerger2` (73!) that ant turns into errors

File: third-party/java/aosp/README.txt (hunk 2 above)
Comment (marcinkwiatkowski) [inline third-party/java/aosp/README.txt]: Buck uses `common-24.2.3`, `sdk-common-24.2.3`, etc. `25.2.0` comes with different api in many places. If we can upgrade to use `25.2.0` everywhere, we can get rid of the whole `aosp` hack and just use prebuild jars

Comment (kageiit) [inline third-party/java/aosp/README.txt]: @dreiss any luck checking this PR out on FB's codebase?

**Comment by sdwilsh** [comment]
A bunch, but not all, builds failed with this: `Error generating manifest file: Manifest merger failed with multiple errors, see logs`.  Unfortunately, I don't see anything in the logs that gives me more details, so it's a bit difficult to tell what's wrong here.

File: src/com/facebook/buck/android/GenerateManifestStep.java [hunk 5]
```diff
@@ -86,35 +83,47 @@ public StepExecutionResult execute(ExecutionContext context) {
...
+          .newMerger(mainManifestFile, logger, ManifestMerger2.MergeType.APPLICATION)
+          .withFeatures(
+              ManifestMerger2.Invoker.Feature.NO_PLACEHOLDER_REPLACEMENT,
+              ManifestMerger2.Invoker.Feature.REMOVE_TOOLS_DECLARATIONS,
+              ManifestMerger2.Invoker.Feature.SKIP_BLAME)
+          .addLibraryManifests(Iterables.toArray(libraryManifestFiles, File.class))
+          .merge();
+      if (mergingReport.getResult().isError()) {
+        mergingReport.log(logger);
+        throw new HumanReadableException("Error generating manifest file");
```
Comment (dreiss) [inline src/com/facebook/buck/android/GenerateManifestStep.java]: This is still showing up with no indication of what the actual error is.  I think we need to display all available diagnostic info in the console to make it easier for engineers (and me) to debug their build failures.

**Comment by dreiss** [comment]
Okay, I'm going to start sorting through the errors now.  The first one is really annoying: when there's a duplicate element conflict, it shows you the full path to the second occurrence, but only the basename of the first.  But all of our basenames are the same!

**Comment by kageiit** [comment]
Any update on this?

**Comment by dreiss** [comment]
Yeah, it breaks some stuff in our builds.  So far, it's all been valid errors in our code, but it's taking my a while to slog through them all.

**Comment by marcinkwiatkowski** [comment]
@dreiss let me know, we could probably support both versions of the merger for some time.

**Comment by dreiss** [comment]
Okay, I found one thing that I think is a legit bug.  If you have a WRITE_EXTERNAL_STORAGE, even if you have a READ_EXTERNAL_STORAGE elsewhere, it adds a READ_EXTERNAL_STORAGE permission, but with an "android:" prefix on the element name, which seems incorrect.  Want to check it out?

**Comment by marcinkwiatkowski** [comment]
So looks like they use "android:" for all implicit permissions. Here's the logic that adds READ_EXTERNAL_STORAGE https://github.com/marcinkwiatkowski/buck/blob/e984badb83ee93f95f52bd2e80629b6e45a9a595/third-party/java/aosp/src/com/android/manifmerger/XmlDocument.java#L489-L489

What's the failure?

**Comment by kageiit** [comment]
> What's the failure?

I think the issue here is that even if `READ_EXTERNAL_STORAGE` is already present elsewhere, it is being re-added as `android: READ_EXTERNAL_STORAGE` from @dreiss 's comment

**Comment by dreiss** [comment]
There's no failure that I know of.  I'm just worried that something is going to go wrong at runtime.

**Comment by kageiit** [comment]
All the apps in the wild using gradle have been on manifest merger 2 for many months now. If there was indeed something to worry about at runtime, it would have been quite obvious :)

You could also sanity test the apk produced by the new merger if needed

**Comment by dreiss** [comment]
There's an internal diff now to upgrade our sdklib to android/sdk-common-25.2.0.jar.  Does that make this easier or harder?

**Comment by marcinkwiatkowski** [comment]
@dreiss doesn't affect this PR directly but we could potentially remove the whole `aosp` and make the whole setup simpler. Are you upgrading all jars in `third-party/java/android`?

**Comment by dreiss** [comment]
I think so.  Not sure about sdklib, but definitely the others.  I'll let you know when it lands.

**Comment by gastoncesarf** [comment]
@marcinkwiatkowski sorry for my question (in the case that is possible) I need to know an approximate date for this. Can you estimate this? Thanks.

**Comment by dreiss** [comment]
Okay, I've narrowed down the problem to addImplicitElements.  Because it thinks that all of our library manifests are targeting sdk 1, it's adding a bunch of (unwanted) stuff to our manifest.  I commented out the call to that method and it seems to be working.  Can you comment it out in the PR, then we can see if we can find a cleaner way to disable these?

**Comment by marcinkwiatkowski** [comment]
@kageiit @dreiss I agree, this is only for a test. 
I think only
```
        if (libraryTargetSdk < 4) {
            addIfAbsent(mergingReport.getActionRecorder(),
                    USES_PERMISSION,
                    permission("WRITE_EXTERNAL_STORAGE"),
                    lowerPriorityDocument.getPackageName() + " has a targetSdkVersion < 4");
            hasWriteToExternalStoragePermission = true;

            addIfAbsent(mergingReport.getActionRecorder(),
                    USES_PERMISSION,
                    permission("READ_PHONE_STATE"),
                    lowerPriorityDocument.getPackageName() + " has a targetSdkVersion < 4");
        }
```
block is problematic for @dreiss

**Comment by dreiss** [comment]
I think the next block (`if (hasWriteToExternalStoragePermission) {`) is also potentially problematic, though we could work around that.  I'm not sure about the last one.

I think what I'd really like (upstream) is that if the library doesn't specify a target, instead of assuming that it is "1" (because, really, what are the odds of that), assume that it doesn't care and set it to the target of the main manifest.

**Comment by marcinkwiatkowski** [comment]
It's a bit of patching but we could do that @dreiss (see latest commit), I think it makes sense. Do most of your libraries don't specify sdk? I updated test data to reflect that.

**Comment by dreiss** [comment]
It looks like this removed both "*Test.java" files.  Was that intentional?  It was messing up our CI with a very confusing error message (which is one reason this has taken so long).

**Comment by marcinkwiatkowski** [comment]
Those were aosp tests, I don't think they need to be run as part of buck pipeline. `GenerateManifestStepTest` is an integration test that covers our use of the merger.

**Comment by dreiss** [comment]
Looks like another failure has been introduced on our side: https://gist.github.com/dreiss/f22391e27e0c65c53c8684781d6591f4

Code looks pretty reasonable to me.  Suggestions?  Maybe create a subclass to use as the second receiver?

**Comment by marcinkwiatkowski** [comment]
Yep, subclass should be good.

**Comment by kageiit** [comment]
Subclass should work fine

**Comment by dreiss** [comment]
This had to be reverted.  I'm looking into it.

**Comment by dreiss** [comment]
Looks like the old version was (somehow) ensuring that all activity-alias elements came after the activities they referenced, but the new one does not, resulting in some `<activity-alias> target activity [...] not found in manifest`.  I guess I'm going to have to track down cases where this happens in our codebase and reorder them manually.  I'm not sure what will happen if the activity and the alias appear in separate files.

I'm open to other suggestions if people have them.

**Comment by marcinkwiatkowski** [comment]
@dreiss I found the diff changed it in old version https://github.com/facebook/buck/commit/2d63b10fe12522283ddb4a5d62cd22dbf8ba0e8a
I think instead of patching the merger, the reordering (if needed) should be done as a separate step on a merged document.

**Comment by dreiss** [comment]
idk, this seems like a bug in the merger.  Why would you require the user to run a separate postprocessing step?

**Comment by dreiss** [comment]
Yep, I found a case in our codebase where the activity and the alias are in separate files, so there's no way to fix this short of postprocessing.  I can do it for Facebook, but this doesn't seem like a reasonable thing for us to ask every Buck user to do.  I'd much rather the merger just use a safe ordering.

**Pull request closed without merging**